x/bank/keeper.go

```diff
@@ -246,3 +286,56 @@ func inputOutputCoins(ctx sdk.Context, am auth.AccountKeeper, inputs []Input, ou
 
 	return allTags, nil
 }
+
+func delegateCoins(
+	ctx sdk.Context, ak auth.AccountKeeper, addr sdk.AccAddress, amt sdk.Coins,
+) (sdk.Tags, sdk.Error) {
+
+	ctx.GasMeter().ConsumeGas(costSubtractCoins, "delegateCoins")
+
+	oldCoins := getCoins(ctx, ak, addr)
+	newCoins := oldCoins.Minus(amt)
+
+	if !newCoins.IsNotNegative() {
+		return nil, sdk.ErrInsufficientCoins(fmt.Sprintf("%s < %s", oldCoins, amt))
+	}
+
+	va, ok := ak.GetAccount(ctx, addr).(auth.VestingAccount)
+	if ok {
+		blockTime := ctx.BlockHeader().Time
+
+		va.TrackDelegation(blockTime, amt)
+		ak.SetAccount(ctx, va)
+	} else {
+		if err := setCoins(ctx, ak, addr, newCoins); err != nil {
+			return nil, err
+		}
+	}
+
+	return sdk.NewTags("action", []byte("delegateCoins")), nil
```

Comment:
can we add the `accAddress` on the tags here? Otherwise it will only have the action tag which is not very useful alone... Also can we move the tags into a `tags/tags.go` file ?

Reply:
Agreed. Updated.